ChronoStore/src/storyreader.cpp

```diff
@@ -0,0 +1,320 @@
+            chunkEnd = currentChunk - 1;
+        }
+        else {
+            chunkStart = currentChunk + 1;
+            ansChunk = currentChunk;
+        }
+    }
+    return ansChunk;
+}
+
+// Search last chunk to retrieve
+int64_t searchLastChunk(ChunkAttr* chunkData, uint64_t timestampToSearch, int64_t chunkEnd, int64_t chunkStart) {
+    int64_t currentChunk = INT_MIN;
+    if(chunkEnd == 0)    return currentChunk;
+
+    while (chunkStart <= chunkEnd) {
+        currentChunk = chunkStart + (chunkEnd - chunkStart) / 2;
+        if (chunkData[currentChunk].endTimeStamp == timestampToSearch) {
+            return currentChunk;
+        }
+        else if (chunkData[currentChunk].endTimeStamp > timestampToSearch) {
+            chunkEnd = currentChunk - 1;
+        }
+        else {
+            chunkStart = currentChunk + 1;
+        }
+    }
+    return chunkStart;
+}
+/**
+ * Read data from story dataset
+ * @param range: Range interval t1:t2 to read from.
+ * @param STORY: story name
+ * @param CHRONICLE: chronicle name
+ * @return: struct DatasetReader ->{0,EventData} (EventData consist of the read values) if successful, else \n
+ *          return {-1,EventData}   (Empty EventData) if failed.
+ */
+DatasetReader storyreader::readFromDataset(std::pair<uint64_t, uint64_t> range, const char* STORY, const char* CHRONICLE){
+    // Disable automatic printing of HDF5 error stack
+    if(!DEBUG){
+        H5Eset_auto(H5E_DEFAULT, NULL, NULL);
+    }
+    std::vector<Event> resultVector;
+    DatasetReader dr(-1, resultVector);
+    hid_t   chronicle, story, storyDatasetSpace, eventType;
+    herr_t  status;
+    ChunkAttr* attributeDataBuffer;
+    Event* eventBuffer;
+
+    /**
+     * Read data from the story
+     */
+    chronicle = H5Fopen(CHRONICLE, H5F_ACC_RDONLY, H5P_DEFAULT);
+    story = H5Dopen2(chronicle, STORY, H5P_DEFAULT);
+    storyDatasetSpace  = H5Dget_space(story);
+
+    if(chronicle < 0 || story < 0 || storyDatasetSpace < 0){
+        if(storyDatasetSpace >= 0)  H5Sclose(storyDatasetSpace);
+        if(story >= 0)  H5Dclose(story);
+        if(chronicle >= 0)  H5Fclose(chronicle);
+        return dr;
+    }
+
+    /**
+     * Open the attribute chunkmetadata
+     * Get the attribute type and storyDatasetSpace
+     * Get the number of DATASET_RANK of the storyDatasetSpace
+     * Get the story creation property list for chunksize information
+     * Read the attribute data into the buffer
+     */
+    hid_t attributeId = H5Aopen(story, ATTRIBUTE_CHUNKMETADATA, H5P_DEFAULT);
+
+    hid_t attributeType = H5Aget_type(attributeId);
+    hid_t attributeDataSpace = H5Aget_space(attributeId);
+
+    hsize_t attributeDims = H5Sget_simple_extent_npoints(attributeDataSpace);
+
+    attributeDataBuffer = (ChunkAttr *) malloc(attributeDims * sizeof(ChunkAttr));
+    herr_t readChunkAttributeErr = H5Aread(attributeId, attributeType, attributeDataBuffer);
```

Comment:
Since we are using STL, may be you should use STL vector here and use new to allocate it on the heap.

Reply:
Done